src/CaptureService/CaptureServiceUtils.cpp

```diff
@@ -0,0 +1,34 @@
+// Copyright (c) 2022 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "CaptureService/CaptureServiceUtils.h"
+
+#include "OrbitBase/Logging.h"
+
+using orbit_grpc_protos::CaptureRequest;
+using orbit_grpc_protos::CaptureResponse;
+
+namespace orbit_capture_service {
+
+CaptureRequest WaitForStartCaptureRequestFromClient(
+    grpc::ServerReaderWriter<CaptureResponse, CaptureRequest>* reader_writer) {
+  CaptureRequest request;
+  // This call is blocking.
+  reader_writer->Read(&request);
+
+  ORBIT_LOG("Read CaptureRequest from Capture's gRPC stream: starting capture");
+  return request;
+}
+
+void WaitForStopCaptureRequestFromClient(
+    grpc::ServerReaderWriter<CaptureResponse, CaptureRequest>* reader_writer) {
+  CaptureRequest request;
+  // The client asks for the capture to be stopped by calling WritesDone. At that point, this call
+  // to Read will return false. In the meantime, it blocks if no message is received.
+  while (reader_writer->Read(&request)) {
+  }
+  ORBIT_LOG("Client finished writing on Capture's gRPC stream: stopping capture");
+}
```

Comment:
The other design question I have (again, I miss the full picture) is related to what we had briefly discussed: Was a gRPC-free interface mimicking this behavior of `grpc::ServerReaderWriter` considered harder to follow than the proposed refactoring?

Reply:
Thanks for the suggestion!

Not sure whether I'm fully understand the purpose of mimicking the behavior of `grpc::ServerReaderWriter`. For the cloud collector, we will not uses the `Capture` rpc to start taking a capture. When a capture is requested by the client, guest orc will create a cloud collector instance and the cloud collector will start taking a capture immediately when other information is ready.
So the cloud collector itself does not need to wait for a start capture command.

Reply:
My general point is: Would there be a way to have a single gRPC-free capture method with the gRPC details used to start and stop the capture abstracted away in some interface?

> Not sure whether I'm fully understand the purpose of mimicking the behavior of grpc::ServerReaderWriter.

The purpose would be that you only have one capture method that doesn't depend on gRPC and you wouldn't need to extract functions to reuse in another capture method as the two methods could end up out of sync.

> the cloud collector will start taking a capture immediately when other information is ready.
> So the cloud collector itself does not need to wait for a start capture command.

A possibility would be that that method of the interface (renamed to not include a reference to "client") returns immediately. At some point we might want to add capture options to the cloud collector and that method would return the capture options (which is basically what happens now, `CaptureRequest` only contains the `CaptureOptions`).

Reply:
Thanks for the explanation!

> Would there be a way to have a single gRPC-free capture method with the gRPC details used to start and stop the capture abstracted away in some interface?

Yes. I thought about this in a deprecated refactoring plan in the shared drawing where we also propose to have a `LinuxCapture` which provides a  gRPC-free `Capture` method to hold all the logic in `LinuxCaptureService::Capture`.

But that one was deprecated as I introduced `void*` in that design. It would be great if you can provide some suggestion on that design.